jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp

```diff
@@ -286,3 +288,84 @@ SCENARIO(
     }
   }
 }
+
+/// Allow access to private methods so that they can be unit tested
+class java_bytecode_convert_method_unit_testt
+{
+public:
+  static exprt
+  convert_aload(const irep_idt &statement, const exprt::operandst &op)
+  {
+    return java_bytecode_convert_methodt::convert_aload(statement, op);
+  }
+};
+
+TEST_CASE(
+  "baload byte array",
+  "[core][java_bytecode][java_bytecode_convert_method][convert_aload]")
+{
+  const typet byte_array_type = java_array_type('b');
+  const symbol_exprt byte_array{"byte_array", byte_array_type};
+  const exprt offset = from_integer(1, java_int_type());
+  const exprt result = java_bytecode_convert_method_unit_testt::convert_aload(
+    "baload", {byte_array, offset});
+
+  // Result is of the form `(int)(*(byte_array->data + offset))`
+  auto query = make_query(result)
+                 .as<typecast_exprt>()[0]
+                 .as<dereference_exprt>()[0]
+                 .as<plus_exprt>();
+  REQUIRE(query[1].get() == offset);
+  auto member = query[0].as<member_exprt>();
+  REQUIRE(member.get().get_component_name() == "data");
+  REQUIRE(member[0].as<dereference_exprt>()[0].get() == byte_array);
+  REQUIRE(result.type() == java_int_type());
+  // byte_array->data has type *byte
+  REQUIRE(member.get().type() == pointer_type(java_byte_type()));
+}
+
+TEST_CASE(
+  "baload boolean array",
+  "[core][java_bytecode][java_bytecode_convert_method][convert_aload]")
+{
+  const typet boolean_array_type = java_array_type('z');
+  const symbol_exprt boolean_array{"boolean_array", boolean_array_type};
+  const exprt offset = from_integer(2, java_int_type());
+  const exprt result = java_bytecode_convert_method_unit_testt::convert_aload(
+    "baload", {boolean_array, offset});
+
+  // Result is of the form `(int)(*(boolean_array->data + offset))`
+  auto query = make_query(result)
+                 .as<typecast_exprt>()[0]
+                 .as<dereference_exprt>()[0]
+                 .as<plus_exprt>();
+  REQUIRE(query[1].get() == offset);
+  REQUIRE(
+    query[0].as<member_exprt>()[0].as<dereference_exprt>()[0].get() ==
+    boolean_array);
+  // boolean_array->data has type *boolean
+  REQUIRE(
+    query[0].as<member_exprt>().get().type() ==
```

Comment:
:question: This seems a little counter-intuitive since JVM stores boolean arrays as byte arrays, is there a good reason why we don't?

Reply:
The type of `boolean` is 8 bits so I think that enough to be compatible with the JVM. The reason we do this is to avoid some unecessary typecasts which then result into byte_extract operations when we attempt to retrieve a boolean from a boolean array that has been cast to a byte array.